cpp/src/rolling/detail/range_comparator_utils.cuh

```diff
@@ -0,0 +1,143 @@
+ *
+ * Note: We use numeric_limits::lowest() instead of min() because for floats, lowest() returns
+ * the smallest finite value, as opposed to min() which returns the smallest _positive_ value.
+ */
+template <typename T, CUDF_ENABLE_IF(cuda::std::numeric_limits<T>::is_signed)>
+__host__ __device__ T subtract_safe(T const& value, T const& delta)
+{
+  if constexpr (std::is_floating_point_v<T>) {
+    if (std::isinf(value) or std::isnan(value)) { return value; }
+  }
+  // delta >= 0;
+  return (value >= 0 || (value - cuda::std::numeric_limits<T>::lowest()) >= delta)
+           ? (value - delta)
+           : cuda::std::numeric_limits<T>::lowest();
+}
+
+/**
+ * @brief Subtract `delta` from value, and cap at numeric_limits::lowest(), for unsigned types.
+ *
+ * Note: We use numeric_limits::lowest() instead of min() because for floats, lowest() returns
+ * the smallest finite value, as opposed to min() which returns the smallest _positive_ value.
+ *
+ * This distinction isn't truly relevant for this overload (because float is signed).
+ * lowest() is kept for uniformity.
+ */
+template <typename T, CUDF_ENABLE_IF(not cuda::std::numeric_limits<T>::is_signed)>
+__host__ __device__ T subtract_safe(T const& value, T const& delta)
+{
+  // delta >= 0;
+  return ((value - cuda::std::numeric_limits<T>::lowest()) >= delta)
+           ? (value - delta)
+           : cuda::std::numeric_limits<T>::lowest();
+}
+
+/**
+ * @brief Comparator for numeric order-by columns, handling floating point NaN values.
+ *
+ * This is required for binary search through sorted vectors that contain NaN values.
+ * With ascending sort, NaN values are stored at the end of the sequence, even
+ * greater than infinity.
+ * But thrust::less would have trouble locating it because:
+ * 1. thrust::less(NaN, 10) returns false
+ * 2. thrust::less(10, NaN) also returns false
+ *
+ * This comparator honours the position of NaN values vis-à-vis non-NaN values.
+ *
+ */
+struct nan_aware_less {
+  template <typename T, CUDF_ENABLE_IF(not cudf::is_floating_point<T>())>
+  __host__ __device__ bool operator()(T const& lhs, T const& rhs) const
+  {
+    return thrust::less<T>{}(lhs, rhs);
+  }
+
+  template <typename T, CUDF_ENABLE_IF(cudf::is_floating_point<T>())>
+  __host__ __device__ bool operator()(T const& lhs, T const& rhs) const
+  {
+    if (std::isnan(lhs)) { return false; }
+    return std::isnan(rhs) or thrust::less<T>{}(lhs, rhs);
+  }
+};
+
+/**
+ * @brief Comparator for numeric order-by columns, handling floating point NaN values. *
+ *
+ * This is required for binary search through sorted vectors that contain NaN values.
+ * With descending sort, NaN values are stored at the beginning of the sequence, even
+ * greater than infinity.
+ * But thrust::greater would have trouble locating it because:
+ * 1. thrust::greater(NaN, 10) returns false
+ * 2. thrust::greater(10, NaN) also returns false
+ *
+ * This comparator honours the position of NaN values vis-à-vis non-NaN values.
```

Comment:
Too close to independence day to allow such language...
```suggestion
 * This comparator honors the position of NaN values vis-à-vis non-NaN values.
```

Reply:
Let the record show that I make this change under protest.